### ISSUE TITLE
Add limitation for retrying mutations

### ISSUE THREAD

**Issue by Alex-Cheng** [ClickHouse/ClickHouse#55777]
**Use case**

When mutations are failed, CH will retry it infinitely until success. But for some cases, the mutations will fail forever. We need an approach to limit the retry attempts and avoid infinite mutation execution. 

**Describe the solution you'd like**

Candidates solutions:
1. Ada a Setting to turn on/off the retry mechanism for mutation;
2. Add a Setting to limit the retry times for failed mutations, failed mutations whose retry times exceeds the limit will be forcely cancelled;
3. Add a Setting to set the time to live(TTL) for unfinished mutations, failed mutations whose life time exceeds the TTL limit will be forcely cancelled.

Any ideas?

**Comment by Alex-Cheng** [comment]
@alexey-milovidov  shall we add this feature? I feel it is useful and actually my team requires it.

**Comment by alexey-milovidov** [comment]
This requires further discussion as it may lead to trouble.

**Comment by Alex-Cheng** [comment]
> This requires further discussion as it may lead to trouble.

In my scenario, the failed mutations keep retrying infinitely and cost a lot of resource. Each time I have to kill mutations manually. It is not easy to do manual operation in customer's environment. So in my opinion, infinite retrying lead to trouble for me.

**Comment by Alex-Cheng** [comment]
@alexey-milovidov  if there is not such a feature available, we decide to implement it ourselves and contribute to the community. Don't hesitate to tell me if you have any comments on it . Thank you.

**Comment by frinkr** [comment]
> This requires further discussion as it may lead to trouble.

Could you be more specific? How about a query level setting, so users can decide to retry or not for each query?

**Comment by Algunenano** [comment]
A couple of scenarios to contemplate:

Scenario 1
* All the parts are in version 5.
* A new mutation arrives. The mutation is set to transform parts from v5 to v6.
* The mutation is completely wrong and none of the parts can be mutated.
* You stop retrying the mutation.
* Old parts are in v5, new parts (the ones inserted after the mutation) are in v6.
* A new mutation arrives (v6 to v7). How is the mutation executed? Should v5->v6 be ignored and do v5->v7 as if it was v6 to v7?

Scenario 2: Same but the mutation but it works for some parts and fails for others:
* All the parts are in version 5.
* A new mutation arrives. The mutation is set to transform parts from v5 to v6.
* The mutation is partially wrong (or it fails sometimes because of OOM for example).
* Some parts are in v5 and some in v6.
* You stop retrying the mutation. Now you have parts in different versions, which means that they can't be merged together.
* A new mutation arrives (v6->v7). How do you execute this one? Do you keep data some data in v5 forever and keep migrating only some parts? How many bugs will you face until you decide that having many metadata versions at the same time is pain?

Basically what I mean with this is that ClickHouse's mutation design is not prepared to just ignore old mutations. Each mutation evolves a table, and CH relies on tables being consistent. If you stop ignoring / retrying mutations then you need to rethink how to manage table evolution in a different way (Do I hear "transactions"?)

**Comment by Alex-Cheng** [comment]
Yes, it is transaction! We need transaction. In my case, the mutation dependent on other table which has been deleted, thus the mutation will keep failing forever. In this case, we need to rollback and cancel the mutation.

Thank you.

BTW, "How many bugs will you face until you decide that having many metadata versions at the same time is pain?" is similar to a famous lyrics in which I love the song :-)

**Comment by Alex-Cheng** [comment]
As there is a feature request defined, I closed this issue.
